Code/Framework/AzCore/AzCore/DOM/DomValue.cpp

```diff
@@ -0,0 +1,1241 @@
+            return aznumeric_cast<uint64_t>(AZStd::get<int64_t>(m_value));
+        case 2: // uint64_t
+            return AZStd::get<uint64_t>(m_value);
+        case 3: // double
+            return aznumeric_cast<uint64_t>(AZStd::get<double>(m_value));
+        }
+        AZ_Assert(false, "AZ::Dom::Value: Called GetInt on a non-numeric type");
+        return {};
+    }
+
+    void Value::SetUint64(uint64_t value)
+    {
+        m_value = value;
+    }
+
+    uint32_t Value::GetUint32() const
+    {
+        return aznumeric_cast<uint32_t>(GetUint64());
+    }
+
+    void Value::SetUint32(uint32_t value)
+    {
+        m_value = aznumeric_cast<uint64_t>(value);
+    }
+
+    bool Value::GetBool() const
+    {
+        if (IsBool())
+        {
+            return AZStd::get<bool>(m_value);
+        }
+        AZ_Assert(false, "AZ::Dom::Value: Called GetBool on a non-bool type");
+        return {};
+    }
+
+    void Value::SetBool(bool value)
+    {
+        m_value = value;
+    }
+
+    double Value::GetDouble() const
+    {
+        switch (m_value.index())
+        {
+        case 1: // int64_t
+            return aznumeric_cast<double>(AZStd::get<int64_t>(m_value));
+        case 2: // uint64_t
+            return aznumeric_cast<double>(AZStd::get<uint64_t>(m_value));
+        case 3: // double
+            return AZStd::get<double>(m_value);
+        }
+        AZ_Assert(false, "AZ::Dom::Value: Called GetInt on a non-numeric type");
+        return {};
+    }
+
+    void Value::SetDouble(double value)
+    {
+        m_value = value;
+    }
+
+    float Value::GetFloat() const
+    {
+        return aznumeric_cast<float>(GetDouble());
+    }
+
+    void Value::SetFloat(float value)
+    {
+        m_value = aznumeric_cast<double>(value);
+    }
+
+    void Value::SetString(AZStd::shared_ptr<const AZStd::string> string)
+    {
+        m_value = string;
+    }
+
+    AZStd::string_view Value::GetString() const
+    {
+        switch (m_value.index())
+        {
+        case 5: // AZStd::string_view
+            return AZStd::get<AZStd::string_view>(m_value);
+        case 6: // AZStd::shared_ptr<const AZStd::string>
+            return *AZStd::get<AZStd::shared_ptr<const AZStd::string>>(m_value);
+        case 7: // ShortStringType
+            {
+                const ShortStringType& ShortString = AZStd::get<ShortStringType>(m_value);
+                return { ShortString.m_data.data(), ShortString.m_size };
+            }
+        }
+        AZ_Assert(false, "AZ::Dom::Value: Called GetString on a non-string type");
+        return {};
+    }
+
+    size_t Value::GetStringLength() const
+    {
+        return GetString().size();
+    }
+
+    void Value::SetString(AZStd::string_view value)
+    {
+        if (value.size() <= ShortStringSize)
+        {
+            ShortStringType buffer;
+            buffer.m_size = value.size();
+            memcpy(buffer.m_data.data(), value.data(), buffer.m_size);
+            m_value = buffer;
```

Comment:
An `AZStd::fixed_string<16>` can just be be used here.

Reply:
@lumberyard-employee-dm Isn't that size going to be the same or close to the small string optimization in `AZStd::string`? Does there even need to be a separate small string optimization in that case?

Reply:
Did a pass on the string storage semantics, thanks to you and @AMZN-koppersr for pointing out that I was being silly and double heap allocating by using `AZStd::string`, storage clases are now either `AZStd::string_view` (ref with no lifecycle guarantee), `AZStd::fixed_string<14>` (small string), or `AZStd::shared_ptr<AZStd::vector<char>>` (the previous shared lifecycle string route, but just using a vector). I haven't seen any hot spots with this approach when profiling at least.